server/etcdserver/apply/apply_auth_test.go

```diff
@@ -15,47 +15,791 @@
+	require.NoError(t, err)
+	authStore := auth.NewAuthStore(
+		lg,
+		schema.NewAuthBackend(lg, be),
+		tp,
+		bcrypt.DefaultCost,
+	)
+	consistentIndex := cindex.NewConsistentIndex(be)
+	return newAuthApplierV3(
+		authStore,
+		newApplierV3Backend(
+			lg,
+			kv,
+			alarmStore,
+			authStore,
+			lessor,
+			cluster,
+			&fakeRaftStatusGetter{},
+			&fakeSnapshotServer{},
+			consistentIndex,
+			false,
+		),
+		lessor)
+}
+
+const (
+	userRoot      = "root"
+	roleRoot      = "root"
+	userReadOnly  = "user_read_only"
+	roleReadOnly  = "role_read_only"
+	userWriteOnly = "user_write_only"
+	roleWriteOnly = "role_write_only"
+
+	key             = "key"
+	rangeEnd        = "rangeEnd"
+	keyOutsideRange = "rangeEnd_outside"
+
+	LeaseId = 1
+)
+
+func mustCreateRolesAndEnableAuth(t *testing.T, authApplier *authApplierV3) {
+	_, err := authApplier.UserAdd(&pb.AuthUserAddRequest{Name: userRoot, Options: &authpb.UserAddOptions{NoPassword: true}})
+	require.NoError(t, err)
+	_, err = authApplier.RoleAdd(&pb.AuthRoleAddRequest{Name: userRoot})
+	require.NoError(t, err)
+	_, err = authApplier.UserGrantRole(&pb.AuthUserGrantRoleRequest{User: userRoot, Role: roleRoot})
+	require.NoError(t, err)
+
+	_, err = authApplier.UserAdd(&pb.AuthUserAddRequest{Name: userReadOnly, Options: &authpb.UserAddOptions{NoPassword: true}})
+	require.NoError(t, err)
+	_, err = authApplier.RoleAdd(&pb.AuthRoleAddRequest{Name: roleReadOnly})
+	require.NoError(t, err)
+	_, err = authApplier.UserGrantRole(&pb.AuthUserGrantRoleRequest{User: userReadOnly, Role: roleReadOnly})
+	require.NoError(t, err)
+	_, err = authApplier.RoleGrantPermission(&pb.AuthRoleGrantPermissionRequest{Name: roleReadOnly, Perm: &authpb.Permission{
+		PermType: authpb.READ,
+		Key:      []byte(key),
+		RangeEnd: []byte(rangeEnd),
+	}})
+	require.NoError(t, err)
 
-	b, _ := betesting.NewDefaultTmpBackend(t)
-	defer betesting.Close(t, b)
+	_, err = authApplier.UserAdd(&pb.AuthUserAddRequest{Name: userWriteOnly, Options: &authpb.UserAddOptions{NoPassword: true}})
+	require.NoError(t, err)
+	_, err = authApplier.RoleAdd(&pb.AuthRoleAddRequest{Name: roleWriteOnly})
+	require.NoError(t, err)
+	_, err = authApplier.UserGrantRole(&pb.AuthUserGrantRoleRequest{User: userWriteOnly, Role: roleWriteOnly})
+	require.NoError(t, err)
+	_, err = authApplier.RoleGrantPermission(&pb.AuthRoleGrantPermissionRequest{Name: roleWriteOnly, Perm: &authpb.Permission{
+		PermType: authpb.WRITE,
+		Key:      []byte(key),
+		RangeEnd: []byte(rangeEnd),
+	}})
+	require.NoError(t, err)
 
-	simpleTokenTTLDefault := 300 * time.Second
-	tokenTypeSimple := "simple"
-	dummyIndexWaiter := func(index uint64) <-chan struct{} {
-		ch := make(chan struct{}, 1)
-		go func() {
-			ch <- struct{}{}
-		}()
-		return ch
+	_, err = authApplier.AuthEnable()
+	require.NoError(t, err)
+}
+
+// SetAuthInfo manually sets the authInfo of the applier. In reality, authInfo is filled before Apply()
+func SetAuthInfo(authApplier *authApplierV3, userName string) {
```

Comment:
Why it's an exported function?
```suggestion
func setAuthInfo(authApplier *authApplierV3, userName string) {
```